File: core-extra/arazzo-parser/src/main/java/org/evomaster/arazzo/deserializer/CriterionTypeDeserializer.java
package org.evomaster.arazzo.deserializer;

import com.fasterxml.jackson.core.JacksonException;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonDeserializer;
import com.fasterxml.jackson.databind.JsonNode;
import org.evomaster.arazzo.models.domain.CriterionExpression;
import org.evomaster.arazzo.models.domain.CriterionType;

import java.io.IOException;

/**
 * Custom Jackson deserializer for {@link CriterionType}.
 * It resolves dynamic JSON payloads by mapping plain text to a {@link CriterionType.Simple},
 * and a JSON object to a {@link CriterionType.Complex}.
 */
public class CriterionTypeDeserializer extends JsonDeserializer<CriterionType> {

    @Override
    public CriterionType deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException, JacksonException {
        JsonNode node = jsonParser.getCodec().readTree(jsonParser);

        if (node.isTextual()) {
            return new CriterionType.Simple(node.asText());
        } else if (node.isObject()) {
            CriterionExpression complex = jsonParser.getCodec().treeToValue(node, CriterionExpression.class);
            return new CriterionType.Complex(complex);
        } else {
            throw new IllegalArgumentException("Arazzo Parsing Error: Invalid " + node.getNodeType() + ". Expected string or Criterion Expression");
        }

    }

}

File: core/src/main/kotlin/org/evomaster/core/problem/rest/arazzo/deserializer/CriterionTypeDeserializer.kt
package org.evomaster.core.problem.rest.arazzo.deserializer

import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.JsonDeserializer
import com.fasterxml.jackson.databind.JsonNode
import org.evomaster.core.problem.rest.arazzo.models.CriterionExpression
import org.evomaster.core.problem.rest.arazzo.models.CriterionType

class CriterionTypeDeserializer : JsonDeserializer<CriterionType>() {
    override fun deserialize(
        p0: JsonParser,
        p1: DeserializationContext
    ): CriterionType {
        val node: JsonNode = p0.codec.readTree(p0)
        return when {
            node.isTextual ->  CriterionType.Simple(node.asText())
            node.isObject -> {
                val complexObj = p0.codec.treeToValue(node, CriterionExpression::class.java)
                CriterionType.Complex(complexObj)
            }
            else -> throw IllegalArgumentException("Arazzo Parsing Error: Invalid ${node.nodeType}. Expected string or Criterion Expression")
        }
    }
}